### PR TITLE
RAC-45: Create/update quantified associations in the External API

### PR DESCRIPTION
**Pull request by tseho**
**Description (for Contributor and Core Developer)**

- Add a structure validator on quantified associations using the same method as we already do for the associations
- Change QuantifiedAssociationsFieldSetter to merge by keys (expected behavior for partial updates)
- End2End tests on the create/update/delete of quantified associations. (including validation)

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Setter/QuantifiedAssociationsFieldSetter.php [hunk 1]
```diff
@@ -26,14 +27,81 @@ class QuantifiedAssociationsFieldSetter extends AbstractFieldSetter
...
+     *     },
+     *     "PRODUCTSET_B": {
+     *         "products": [
+     *              {"identifier": "AKN_TS1", "quantity": 2}
+     *         ],
+     *         "product_models": [],
+     *     },
+     * }
+     */
+    private function mergeByKeys(array $quantifiedAssociations1, array $quantifiedAssociations2): array
```
Comment (juliensnz) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Setter/QuantifiedAssociationsFieldSetter.php:92]: not a big fan of "quantifiedAssociations1" and "quantifiedAssociations2" I would prefer "firstQuantifiedAssociations" even if it's not perfect.

Comment (juliensnz) [reply src/Akeneo/Pim/Enrichment/Component/Product/Updater/Setter/QuantifiedAssociationsFieldSetter.php:92]: it's not a big problem though, you can keep it that way without any problem ^^

Comment (tseho) [reply src/Akeneo/Pim/Enrichment/Component/Product/Updater/Setter/QuantifiedAssociationsFieldSetter.php:92]: I'm not a big fan of it neither, and we already discussed about it with @SamirBoulil when we were working on the merger : https://github.com/akeneo/pim-community-dev/blob/MET-198/src/Akeneo/Pim/Enrichment/Component/Product/QuantifiedAssociation/QuantifiedAssociationsMerger.php#L45

So I just wanted to stay consistent :)

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/ProductModelUpdater.php [hunk 2]
```diff
@@ -45,14 +46,18 @@ class ProductModelUpdater implements ObjectUpdaterInterface
     /** @var QuantifiedAssociationsFromAncestorsFilter */
     private $quantifiedAssociationsFromAncestorsFilter;
 
+    /** @var QuantifiedAssociationsStructureValidatorInterface */
+    private $quantifiedAssociationsStructureValidator;
+
     /**
-     * @param PropertySetterInterface               $propertySetter
-     * @param ObjectUpdaterInterface                $valuesUpdater
+     * @param PropertySetterInterface $propertySetter
```
Comment (ValentinMumble) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/ProductModelUpdater.php:53]: I think we should get rid of this now useless phpdoc 🤔

File: tests/back/Pim/Enrichment/EndToEnd/Product/ProductModel/ExternalApi/AbstractProductModelTestCase.php [hunk 3]
```diff
@@ -125,6 +126,21 @@ protected function getStandardizedProductModels()
         return $standardizedProductModels;
     }
 
+    /**
+     * @param array  $expectedProductModel normalized data of the product model that should be created
+     * @param string $identifier           identifier of the product that should be created
+     */
+    protected function assertSameProductModels(array $expectedProductModel, $identifier)
```
Comment (StevenVAIDIE) [inline tests/back/Pim/Enrichment/EndToEnd/Product/ProductModel/ExternalApi/AbstractProductModelTestCase.php:133]: Maybe rename $identifier to $code even if we keep call to findOneByIdentifier ?

```suggestion
     * @param array  $expectedProductModel normalized data of the product model that should be 
     * @param string $code           code of the product model that should be compared
     */
    protected function assertSameProductModels(array $expectedProductModel, string $identifier): void
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Validator/QuantifiedAssociationsStructureValidator.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
+<?php
+
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Validator/QuantifiedAssociationsStructureValidator.php:2]:
```suggestion
declare(strict_types=1);
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Validator/QuantifiedAssociationsStructureValidatorInterface.php [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<?php
+
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Validator/QuantifiedAssociationsStructureValidatorInterface.php:2]:
```suggestion
declare(strict_types=1);
```

Comment (tseho) [reply src/Akeneo/Pim/Enrichment/Component/Product/Updater/Validator/QuantifiedAssociationsStructureValidatorInterface.php:2]: strict types in an interface ?

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/ProductModelUpdater.php (hunk 2 above)
Comment (tseho) [reply src/Akeneo/Pim/Enrichment/Component/Product/Updater/ProductModelUpdater.php:53]: For new classes, I do agree, for "legacy" classes, I try to touch as little lines as possible.

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Validator/QuantifiedAssociationsStructureValidatorInterface.php (hunk 5 above)
Comment (StevenVAIDIE) [reply src/Akeneo/Pim/Enrichment/Component/Product/Updater/Validator/QuantifiedAssociationsStructureValidatorInterface.php:2]: Yes in effect it's useless

**Pull request merged**